File: golang/src/detectors/weak-random-number-generation/WeakRandomNumberGeneration.go
package main

import (
	crand "crypto/rand"
	"math/big"
	mrand "math/rand"
	"time"
)

var charset = []byte("abcdeABCDE0123456789")

// {fact rule=weak-random-number-generation@v1.0 defects=1}
func weakRandomNumberGenerationNoncompliant(length int) string {
	r := mrand.New(mrand.NewSource(time.Now().UnixNano()))
    result := make([]byte, length)
    for i := range result {
        // Noncompliant: `math.rand` is used for generating random number.
        result[i] = charset[r.Intn(len(charset))]
    }
    return string(result)
}
// {/fact}

// {fact rule=weak-random-number-generation@v1.0 defects=0}
func weakRandomNumberGenerationCompliant(length int) string  {
	result := make([]byte, length)
	max := big.NewInt(int64(len(charset)))
	for i := range result {
		// Noncompliant: `math.rand` is used for generating random number.
		n, _ := crand.Int(crand.Reader, max)
		result[i] = charset[n.Int64()]
	}
	return string(result)
}
// {/fact}

File: csharp/src/detectors/weak-random-number-generation/WeakRandomNumberGeneration.cs
using System.Security.Cryptography;

namespace Cryptography
{
    class WeakRandomNumberGeneration
    {
        // {fact rule=weak-random-number-generation@v1.0 defects=1}
        public void WeakRandomNumberGenerationNoncompliant()
        {
            var randomnumber = new System.Random();
            byte[] key = new byte[16];
            randomnumber.NextBytes(key);
            // Noncompliant: An insecure random number generator (RNG) is used to create cryptographic key.
            var c = new AesGcm(key);
        }
        // {/fact}
        
        // {fact rule=weak-random-number-generation@v1.0 defects=0}
        public void WeakRandomNumberGenerationCompliant()
        {
            var randomnumber = System.Security.Cryptography.RandomNumberGenerator.Create();
            byte[] key = new byte[16];
            randomnumber.GetBytes(key);
            // Compliant: Secure random number generator (RNG) is used to create cryptographic key.
            var c = new AesGcm(key);
        }
        // {/fact}
    }
}
